c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -1763,6 +2126,143 @@ class DeviceCachingAllocator {
     return blocks;
   }
 
+  // returns the smallest possible address in any segment
+  // where there is enough free address space to fit size
+  // may be composed of free and unmapped segments
+  Block* find_expandable_block(
+      int device,
+      cudaStream_t stream,
+      BlockPool* pool,
+      size_t size) {
+    Block key(device, stream, 0);
+
+    auto allocateable = [](Block* b) {
+      return b && !b->allocated && b->event_count == 0 &&
+          b->stream_uses.empty();
+    };
+    auto has_available_address_space = [&](Block* b) {
+      size_t bytes = 0;
+      while (bytes < size && allocateable(b)) {
+        bytes += b->size;
+        b = b->next;
+      }
+      return bytes >= size;
+    };
+    for (auto it = pool->unmapped.lower_bound(&key);
+         it != pool->unmapped.end() && (*it)->stream == stream;
+         ++it) {
+      Block* c = *it;
+      if (allocateable(c->prev)) {
```

Comment:
Just wondering about this, why does the `pool->unmapped.lower_bound` search not return `c-prev` directly and we have to take the iterator and then access the `prev` ?

Reply:
This bit is a little subtle. We want to find the lowest address at which there is enough space to fit the allocation. We know already it cannot fit into free block because we wouldn't get to this code, but `unmapped` only tracks unmapped blocks (unallocated and unmapped). There might be a free block right before it. We do not have to keep searching beyond this one free block, but any unmapped blocks would have already been considered earlier in the search.